src/clientiface.cpp

```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+				if (m_blocks_updated_since_last_send.count(wms) == 0)
+					continue;
 			}
 
-			/*
-				Don't send already sent blocks
-			*/
-			if (m_blocks_sent.find(p) != m_blocks_sent.end())
-				continue;
+			// If the MapBlock is not loaded, it will be queued to be loaded or
+			// generated. Otherwise it will be added to 'dest'.
 
-			/*
-				Check if map has this block
-			*/
-			MapBlock *block = env->getMap().getBlockNoCreateNoEx(p);
+			const bool allow_generate = true;
 
-			bool block_not_found = false;
-			if (block) {
+			MapBlock *block = m_env->getMap().getBlockNoCreateNoEx(wms.p);
+
+			bool surely_not_found_on_disk = false;
+			bool emerge_required = false;
+			if(block != NULL)
+			{
 				// Reset usage timer, this block will be of use in the future.
 				block->resetUsageTimer();
 
-				// Check whether the block exists (with data)
-				if (block->isDummy() || !block->isGenerated())
-					block_not_found = true;
+				// Block is dummy if data doesn't exist.
+				// It means it has been not found from disk and not generated
+				if(block->isDummy())
+					surely_not_found_on_disk = true;
 
-				/*
-					If block is not close, don't send it unless it is near
-					ground level.
-
-					Block is near ground level if night-time mesh
-					differs from day-time mesh.
-				*/
-				if (d >= d_opt) {
-					if (!block->getIsUnderground() && !block->getDayNightDiff())
-						continue;
-				}
+				if(block->isGenerated() == false && allow_generate)
+					emerge_required = true;
 
-				if (m_occ_cull && !block_not_found &&
-						env->getMap().isBlockOccluded(block, cam_pos_nodes)) {
-					continue;
-				}
+				// This check is disabled because it mis-guesses sea floors to
+				// not be worth transferring to the client, while they are.
+				/*if (d >= 4 && block->getDayNightDiff() == false)
+					continue;*/
 			}
 
-			/*
-				If block has been marked to not exist on disk (dummy) or is
-				not generated and generating new ones is not wanted, skip block.
-			*/
-			if (!generate && block_not_found) {
-				// get next one.
-				continue;
+			if(allow_generate == false && surely_not_found_on_disk == true) {
+				// NOTE: If we wanted to avoid generating new blocks based on
+				// some criterion, that check woould go here and we would call
+				// 'continue' if the block should not be generated.
+
+				// TODO: There needs to be some new way or limiting which
+				// positions are allowed to be generated, because we aren't
+				// going to look up the player's position and compare it with
+				// max_mapblock_generate_distance anymore. Maybe a configured set
+				// of allowed areas, or maybe a callback to Lua.
+
+				// NOTE: There may need to be a way to tell the client that this
+				// block will not be transferred to it no matter how nicely it
+				// asks. Otherwise the requested send queue is going to get
+				// filled up by these and less important blocks further away
+				// that happen to have been already generated will not transfer.
```

Comment:
so no `continue` here?

also I didn't manage to determine whether these changes reintroduce the following bug (I have a feeling it might): 5430770b6851a18417f6b2629c9167a5aa38baa5